### PR TITLE
Add robustness tests and further model improvements

### PR DESCRIPTION
**Pull request by skaupper**
Use two different custom exception types for errors in the model:
- `CommunicationError`\
   Signals errors caused by the RMI communication itself the cause may be a RemoteException (if the RMI call itself failed), or a TimeoutException in case the maximum number of retries has been exceeded (and the data is still not consistent).
- `ConnectionError`\
  Connecting/reconnecting to the RMI registry failed.

Localize error messages, so they can be output to the user.

### CODE REVIEW

File: src/test/java/at/fhhagenberg/esd/sqe/ws20/gui/ErrorTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    void testEmptyErrorLogAfterStartup() {
+        page.assertEmptyErrorLog();
+    }
+
+    @Test
+    void testNonEmptyErrorLogAfterException() throws RemoteException {
+        when(mockElevator.getClockTick()).thenThrow(RemoteException.class);
+        page.selectElevator(0);
+        page.assertNonEmptyErrorLog();
```
Comment (skaupper) [inline src/test/java/at/fhhagenberg/esd/sqe/ws20/gui/ErrorTest.java:65]: After selecting an elevator (and therefore querying model data) the thrown exception should appear in the error log.

File: src/test/java/at/fhhagenberg/esd/sqe/ws20/gui/ErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    void testRecurringErrors() throws RemoteException {
+        when(mockElevator.getClockTick()).thenThrow(RemoteException.class);
+        page.assertNonEmptyErrorLogWithTimeout();
+        page.clearErrorLog();
+        page.assertNonEmptyErrorLogWithTimeout();
```
Comment (skaupper) [inline src/test/java/at/fhhagenberg/esd/sqe/ws20/gui/ErrorTest.java:74]: The `update()` routine should trigger errors even if the user does not interact with the GUI at all. We may decide to handle recurring `ConnectionError`s differently (in order not to spam the user with the same error over and over again) but that would be a different testcase.

File: src/test/java/at/fhhagenberg/esd/sqe/ws20/gui/ModelUpdateTest.java [hunk 3]
```diff
@@ -44,10 +41,8 @@ public void setup(FxRobot robot) {
     }
 
 
-    @Disabled("The GUI behaviour does not yet match the behaviour described in this test.")
     @Test
     void testTargetFloorUpdating() {
-        // TODO: fix the GUI behaviour for this test to work!
         elevatorRMIMock.setTarget(0, 1);
         elevatorRMIMock.setTarget(1, 2);
```
Comment (skaupper) [inline src/test/java/at/fhhagenberg/esd/sqe/ws20/gui/ModelUpdateTest.java:47]: This test will fails as long as the GUI cannot handle changing target floors.

**Comment by skaupper** [comment]
@pagdot @MADITT 
We have 3 failing (new) test cases atm. Before this PR can be merged the missing features must be implemented and merged to master. Since it is mostly controller related stuff it's probably a topic for @pagdot but idc too much^^

**Comment by pagdot** [comment]
> @pagdot @MADITT
> We have 3 failing (new) test cases atm. Before this PR can be merged the missing features must be implemented and merged to master. Since it is mostly controller related stuff it's probably a topic for @pagdot but idc too much^^

I can take a look into it. When do you need it done?

**Comment by skaupper** [comment]
> > @pagdot @MADITT
> > We have 3 failing (new) test cases atm. Before this PR can be merged the missing features must be implemented and merged to master. Since it is mostly controller related stuff it's probably a topic for @pagdot but idc too much^^
> 
> I can take a look into it. When do you need it done?

According to Moodle the assignment is due tomorrow evening. Even though that date is not that strict I would like to integrate it sometime at wednesday.

**Comment by skaupper** [comment]
> Kudos, SonarCloud Quality Gate passed!
> 
> [Bug](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&metric=new_coverage&view=list) [84.3% Coverage](https://sonarcloud.io/component_measures?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=fhhagenberg-sqe-esd-ws20_elevator-control-center-team-a&pullRequest=7&metric=new_duplicated_lines_density&view=list)

Good enough, I guess.

**Comment by MADITT** [comment]
Looks fine to me.

**Review by MADITT — COMMENTED** [review]
Looks fine to me.

**Pull request merged**